### PR TITLE
[SCFToCalyx] Fix memory load followed by store.

### LINKED ISSUE #2108: [SCFToCalyx] Incorrect assumptions for memory inlining.
Currently, the SCFToCalyx pass assumes that if there is a single load from a given memory, then it can be "inlined."

This means a body with a read and write to the same memory will be inlined. This leads to two individual problems:

1. We probably shouldn't be reading and writing to the same memory within the same group.
2. Address ports are assigned twice; this is illegal.

### PR DESCRIPTION
**Pull request by cgyurgyik**
Closes #2108. This takes the very safe assumption that if *any* memory stores are present, do not throw the load in a combinational group.

Also some random cleanup, and the removal of a dump call.

### CODE REVIEW

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp [hunk 1]
```diff
@@ -790,38 +798,42 @@ class BuildOpGroups : public FuncOpPartialLoweringPattern {
...
                                      memref::LoadOp loadOp) const {
-  auto memoryInterface =
-      getComponentState().getMemoryInterface(loadOp.memref());
-  if (singleLoadFromMemory(loadOp)) {
-    /// Single load from memory; Combinational case - we do not have to consider
-    /// adding registers in front of the memory.
+  Value memref = loadOp.memref();
+  auto memoryInterface = getComponentState().getMemoryInterface(memref);
+  if (noStoresToMemory(memref) && singleLoadFromMemory(memref)) {
+    // Single load from memory; we do not need to write the
```
Comment (mortbopet) [inline lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:804]: Thank you for elaborating this :+1:.

**Review by mortbopet — APPROVED** [review]
Looks good to me!

**Review by mikeurbach — APPROVED** [review]
Makes sense to me.

A side note on this for myself and @jopperm to consider if we are talking about scheduling a computation for Calyx... since some reads may be combinational and some may be registerred, it would be great if we could reflect this when we construct the scheduling problem with Calyx in mind. We could potentially do a similar analysis and use a 0-cycle latency operator type on combinational reads, and a 1-cycle latency operator type on reads we know will be registerred. In my prototyping so far, I've just considered all reads to have 1-cycle latency to be safe.

**Comment by cgyurgyik** [comment]
In general, you're saying give memory operations (load, store) a specific latency for static scheduling ?

**Comment by cgyurgyik** [comment]
Right, so IIUC you're saying the latencies for load/store should be defined per memory primitive.

**Pull request merged**